### ISSUE TITLE
(WIP) SLAs: Support TTLs additionally to interruption

### ISSUE THREAD

**Issue by domdorn** [zio/zio#5592]
This is a WIP as I'm on mobile and the ios app just lost the whole ticket text I wrote, so I will create this ticket and update it till everything is written, then remove the (WIP).

This is a follow up on the following twitter discussion: https://twitter.com/viktorklang/status/1437344391052541952?s=21 

The current way to implement Service Level Agreements (SLAs) in ZIO is by wrapping an effect with e.g. `timeout(1.second)`, which would `interrupt` the request processing fiber when the specified time has passed. 
However, this does (obviously) not work for effects that are marked `uninterruptible` or that call into systems that are not interruptiple, e.g.
- SOAP web services
- Calls to other web services that dont interrupt when the inputstream/socket is closed
- long running JDBC queries/stored procedures
- legacy systems that are integrated using terminal emulation (eg 32/70)
- ...

Another (additional) approach would be to provide means to avoid certain computations/calls altogether, when we know/can derive that we will not finish in time.

For this we would need the following information:
- How much time in total are we allowed to use over all the included effects in this request
- How much of that have we already spent/how much is still available

I'd summarize this info as TTL = Tme to Life, as this is the same concept already found in networking and the term is used there.

With this information, an effect could make the decision if it should run a certain call/code (because it might succeed in time) or give up at this step as it might know already that it will finish too late and the computation is therefor useless altogether.

I think the information "how much time has passed" should be available through the metrics part of ZIO 2. We'd need to make an API to access that. 

The information about "how much time is the whole chain allowed to take" is not available at the moment. We could provide this through the environment, e.g. similar to how we propagate log-contexts or maybe in a similar way we are passing interruptible/uninterruptible (mask) information, thus in a way.

I could imagine a set of operators, e.g. 

```scala
ZIO.withSLA[R,E,A](d: Duration)(e: ZIO[R, E, A]): ZIO[R,E,A] // or ZIO[R with Has[SLA], E, A]
ZIO.attemptSLACond(e: SLA => Boolean)(e: ZIO[R,E,A]): ZIO[R, E,A]
ZIO.attemptSLA(e: SLA => ZIO[R,E,A]): ZIO[R, E, A]
```

maybe with something like 
```scala
final case class SLA(maxDuration: Duration, alreadySpent: Duration) { 
 def willSucceedInTime(d: Duration): Boolean = ??? 
} 
```

Usage then would be something like this:

In my controller (pseudo code):
```scala

case class MyController(service: SomeService) { 

   def route = (get & path("/api/checkReservation") & extractReq[CheckReservationReq]) { req => 
      val serviceReq = toServiceReq(req)
      service.checkReservation(serviceReq).withSLA(1.second).map(r => respond(200, toJson(r)))
   }
...
}
```

and e.g. in one of the servces called:

```scala

case class LufthansaService(client: LufthansaClient) { 
 val medianReponse = 200.millis
 
 def checkReservation(someParams...) = ZIO.attemptSLACond(sla => sla.willSucceedIn(medianResponse)(client.makeReq(...))
 
 ```
 
 or maybe in another one:
 
 ```scala
 case class QuantasService(client: QuantasClient) {
  val medianResponse = 150.millis
  
  def checkReservation(someParams...) = client.makeReq(...).slaCond(sla => sla.willSucceedIn(medianResponse))
  
}
```

or maybe this one:

```scala
case class ThaiService(client: ThaiClient) { 
val medianResponse = 125.millis

 def checkReservation(someParams) = ZIO.getSLA.flatMap(sla => if(sla.willSucceedIn(medianResponse)) { client.call } else { ZIO.interrupt } 
 ```
 
It is probably a lot of work to get this into ZIO. A poor mans approach would be to create a `Ref[SLA]` and put that into the environment and pass it down the effects. This has the disadvantage, that every effect in the chain has to be adjusted. If possible I would make this a part of core zio and handle failures in the SLA the same way we do interruptions.

**Comment by adamgfraser** [comment]
While this is definitely an interesting design principle, I tend to think that it is best addressed at higher levels using the existing tools provided by ZIO.

If we can interrupt a service that is obviously the best case scenario. If we can't interrupt the service ourselves, for example because it is occurring on a remote server, then we can transmit to it a time we are willing to wait. At that point it is up to that service what to do with that information. If the remote service is implemented in terms of `ZIO` the remote service could immediately take the client specified time to wait and use it to apply a timeout to its internal computations, for example. Or the remote service could use other logic specific to its implementation to decide what actions to take or to cancel at some time.

Since we don't know the nature of these services I think the best we can do is provide an easy tool to determine "how much time is left" that services can then use to determine their own logic.

This can be done quite easily using existing operators by getting the current time when a request is received and comparing the current time to that at any desired point.

```scala
import zio._

object Example extends ZIOApp {

  def withTimeToWait[R, E, A](timeToWait: Duration)(f: UIO[Duration] => ZIO[R, E, A]): ZIO[R with Has[Clock], E, A] =
    for {
      clock    <- ZIO.service[Clock]
      start    <- clock.instant
      remaining = clock.instant.map(now => Duration.fromInterval(now, start.plus(timeToWait)))
      a        <- f(remaining)
    } yield a

  def run =
    withTimeToWait(5.seconds) { getRemaining =>
      for {
        _         <- ZIO.debug("starting")
        _         <- ZIO.debug("do something").delay(8.seconds)
        remaining <- getRemaining
        _ <- if (remaining <= Duration.Zero) ZIO.debug("ending early")
             else ZIO.debug("doing more work").delay(7.seconds)
      } yield ()
    }
}
```

This could potentially be packaged up in some service but I think since you potentially want to do this at different scopes and most of the logic is in what a particular application does with the information rather than exposing the information this is probably best implemented in user land.

**Comment by domdorn** [comment]
after sleeping over it, I think we don't even need a `withTimeToWait` or `withSLA` "concept". 

The only thing we would need is 
- to actually track time (maybe always or only if there is a `timeout` attached to an effect) and
- a way to extract the passed/remaining time, either to 
  - pass that information to external systems
  - decide to interrupt/stop processing early

**Comment by adamgfraser** [comment]
Yeah it just gives you this `timeRemaining` effect that you can pass around as a value and call as many times as you want to see how much time you still have remaining. Depending on how many layers of your application you wanted to pass it through you could potentially put it in some data structure or something like a `FiberRef` in the environment.
